p2p/protocols/protocol.go

```diff
@@ -225,31 +231,111 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
 	}
 }
 
-// Run starts the forever loop that handles incoming messages
-// called within the p2p.Protocol#Run function
-// the handler argument is a function which is called for each message received
+// Run starts the forever loop that handles incoming messages.
+// The handler argument is a function which is called for each message received
 // from the remote peer, a returned error causes the loop to exit
-// resulting in disconnection
+// resulting in disconnection of the protocol
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	wait := p.run(handler)
+	return wait()
+}
+
+// run receives messages from the peer and dispatches async routines to handle the messages
+func (p *Peer) run(handler func(ctx context.Context, msg interface{}) error) func() error {
+	p.mtx.Lock()
+	p.running = true
+	p.mtx.Unlock()
+
+	errc := make(chan error)
+	go func() {
+		for {
+			msg, err := p.readMsg()
+			if err != nil {
+				select {
+				case errc <- err:
+				default:
+				}
+
+				return
 			}
-			return err
+
+			p.mtx.RLock()
+			// if loop has been stopped, we don't dispatch any more async routines
+			if !p.running {
+				p.mtx.RUnlock()
+				continue
+			}
+
+			p.wg.Add(1)
+			go func() {
+				defer p.wg.Done()
+				err := p.handleMsg(msg, handler)
+				if err != nil {
+					p.mtx.Lock()
+					p.running = false
+					p.mtx.Unlock()
+
+					select {
+					case errc <- err:
+					default:
+					}
+				}
+			}()
+
+			p.mtx.RUnlock()
 		}
+	}()
+
+	return func() error {
+		return <-errc
 	}
 }
 
-// Drop disconnects a peer.
+func (p *Peer) readMsg() (*p2p.Msg, error) {
+	msg, err := p.rw.ReadMsg()
+	if err != nil {
+		if err != io.EOF {
+			metrics.GetOrRegisterCounter("peer.readMsg.error", nil).Inc(1)
+			log.Error("peer.readMsg", "err", err)
+		}
+	}
+	return &msg, err
```

Comment:
p2p.Msg is using value semantics, no need for pointer. This is causing a unneeded dereference in handleMsg.